File: app/src/main/java/com/cimdriver/app/util/TripClassification.kt
package com.cimdriver.app.util

import com.cimdriver.app.data.local.entity.ClassificationRule
import java.util.Calendar

enum class TripCategory {
    BUSINESS,
    PRIVATE,
    COMMUTE
}

object TripClassification {
    const val DYNAMICS_HOME_TO_WORK = "Home To Work"
    const val DYNAMICS_BUSINESS_MEETING = "Business Meeting"
    const val DYNAMICS_PERSONAL = "PERSONAL"

    private val businessTripTypes = setOf(
        "BUSINESS",
        "Business Meeting",
        "Customer Visit",
        "Customer Billable",
        "Commissioned By CIMSOLUTIONS",
        "Exam Course",
        "Car Maintenance"
    )

    fun classify(
        tripType: String?,
        startAddressType: String? = null,
        endAddressType: String? = null,
        startAddress: String? = null,
        endAddress: String? = null,
        defaultCategory: TripCategory = TripCategory.PRIVATE,
        homeWorkAsCommute: Boolean = true,
        customerAsBusiness: Boolean = true,
        rules: List<ClassificationRule> = emptyList()
    ): TripCategory? {
        val isCommuteAddressPair = (startAddressType == "THUIS" && endAddressType == "WERK") ||
            (startAddressType == "WERK" && endAddressType == "THUIS")

        val matchingRule = findMatchingRule(tripType, startAddressType, endAddressType, startAddress, endAddress, rules)
        if (matchingRule != null) {
            return when (matchingRule.category) {
                "BUSINESS" -> TripCategory.BUSINESS
                "COMMUTE" -> TripCategory.COMMUTE
                "PRIVATE" -> TripCategory.PRIVATE
                else -> defaultCategory
            }
        }
        // If rules exist but none matched, fall through to built-in classification logic
        // so that smart defaults (e.g. KLANT→THUIS = BUSINESS) still apply.

        return when {
            homeWorkAsCommute && (isCommuteAddressPair || tripType == "COMMUTE" || tripType == DYNAMICS_HOME_TO_WORK) -> TripCategory.COMMUTE
            customerAsBusiness && (startAddressType == "KLANT" || endAddressType == "KLANT") -> TripCategory.BUSINESS
            tripType in businessTripTypes -> TripCategory.BUSINESS
            tripType == "PRIVATE" || tripType == DYNAMICS_PERSONAL -> TripCategory.PRIVATE
            else -> defaultCategory
        }
    }

    fun resolveTripType(
        tripType: String?,
        startAddressType: String?,
        endAddressType: String?,
        startAddress: String?,
        endAddress: String?,
        rules: List<ClassificationRule>
    ): String? = findMatchingRule(tripType, startAddressType, endAddressType, startAddress, endAddress, rules)?.tripType ?: tripType

    private fun findMatchingRule(
        tripType: String?,
        startAddressType: String?,
        endAddressType: String?,
        startAddress: String?,
        endAddress: String?,
        rules: List<ClassificationRule>
    ): ClassificationRule? = rules.filter { rule ->
        val tripMatches = rule.tripType.isNullOrBlank() || rule.tripType.equals(tripType, ignoreCase = true)
        
        val typeDirectionForward = (rule.startAddressType.isNullOrBlank() || rule.startAddressType.equals(startAddressType, ignoreCase = true)) &&
            (rule.endAddressType.isNullOrBlank() || rule.endAddressType.equals(endAddressType, ignoreCase = true))
            
        val typeDirectionReverse = (rule.startAddressType.isNullOrBlank() || rule.startAddressType.equals(endAddressType, ignoreCase = true)) &&
            (rule.endAddressType.isNullOrBlank() || rule.endAddressType.equals(startAddressType, ignoreCase = true))

        val exactDirectionForward = (!rule.startAddress.isNullOrBlank() && rule.startAddress.equals(startAddress, ignoreCase = true)) &&
            (!rule.endAddress.isNullOrBlank() && rule.endAddress.equals(endAddress, ignoreCase = true))
            
        val exactDirectionReverse = (!rule.startAddress.isNullOrBlank() && rule.startAddress.equals(endAddress, ignoreCase = true)) &&
            (!rule.endAddress.isNullOrBlank() && rule.endAddress.equals(startAddress, ignoreCase = true))
            
        // Rule matches if trip type matches AND (types match OR exact addresses match)
        // If a rule defines startAddress/endAddress, it must match them exactly.
        val matchesTypes = typeDirectionForward || typeDirectionReverse
        val matchesExact = exactDirectionForward || exactDirectionReverse
        
        val usesExactAddresses = !rule.startAddress.isNullOrBlank() && !rule.endAddress.isNullOrBlank()
        
        tripMatches && (if (usesExactAddresses) matchesExact else matchesTypes)
    }.maxByOrNull { rule ->
        listOf(rule.startAddressType, rule.endAddressType, rule.tripType, rule.startAddress, rule.endAddress).count { !it.isNullOrBlank() }
    }

    fun isOutsideWorkHours(timestamp: Long, workDaysStr: String, workStartTime: String, workEndTime: String): Boolean {
        if (workDaysStr.isBlank() || workStartTime.isBlank() || workEndTime.isBlank()) return false
        
        val cal = Calendar.getInstance()
        cal.timeInMillis = timestamp
        
        // Calendar days: 1=Sun, 2=Mon, 3=Tue, 4=Wed, 5=Thu, 6=Fri, 7=Sat
        // Settings workDays format: "1,2,3,4,5" where 1=Mon, 7=Sun
        val calDayOfWeek = cal.get(Calendar.DAY_OF_WEEK)
        // Convert to our format (1=Mon, 7=Sun)
        val ourDayOfWeek = if (calDayOfWeek == Calendar.SUNDAY) 7 else calDayOfWeek - 1
        
        val workDaysList = workDaysStr.split(",").mapNotNull { it.trim().toIntOrNull() }
        
        if (!workDaysList.contains(ourDayOfWeek)) {
            return true // It's outside work days
        }
        
        // Check hours
        try {
            val startParts = workStartTime.split(":")
            val endParts = workEndTime.split(":")
            if (startParts.size >= 2 && endParts.size >= 2) {
                val startHour = startParts[0].toInt()
                val startMin = startParts[1].toInt()
                val endHour = endParts[0].toInt()
                val endMin = endParts[1].toInt()
                
                val currentHour = cal.get(Calendar.HOUR_OF_DAY)
                val currentMin = cal.get(Calendar.MINUTE)
                
                val currentMins = currentHour * 60 + currentMin
                val startMins = startHour * 60 + startMin
                val endMins = endHour * 60 + endMin
                
                if (currentMins < startMins || currentMins > endMins) {
                    return true // Outside work hours
                }
            }
        } catch (e: Exception) {
            // If parsing fails, fall back to false
        }
        return false
    }
}

File: ios/CIMDriver/Utilities/TripClassification.swift
import Foundation

enum TripCategory: String, Codable {
    case business = "BUSINESS"
    case privateTrip = "PRIVATE"
    case commute = "COMMUTE"
}

enum TripClassification {
    static let dynamicsHomeToWork = "Home To Work"
    static let dynamicsBusinessMeeting = "Business Meeting"
    static let dynamicsPersonal = "PERSONAL"

    private static let businessTripTypes: Set<String> = [
        "BUSINESS",
        "Business Meeting",
        "Customer Visit",
        "Customer Billable",
        "Commissioned By CIMSOLUTIONS",
        "Exam Course",
        "Car Maintenance"
    ]

    static func classify(
        tripType: String?,
        startAddressType: String? = nil,
        endAddressType: String? = nil,
        startAddress: String? = nil,
        endAddress: String? = nil,
        defaultCategory: TripCategory = .privateTrip,
        homeWorkAsCommute: Bool = true,
        customerAsBusiness: Bool = true,
        rules: [ClassificationRule] = [],
        timestamp: Date? = nil,
        appSettings: AppSettings? = nil
    ) -> TripCategory {
        let isCommuteAddressPair = (startAddressType == "THUIS" && endAddressType == "WERK") ||
            (startAddressType == "WERK" && endAddressType == "THUIS")
            
        // Outside work hours check first (unless there's an explicit rule)
        let outsideHours = timestamp != nil ? isOutsideWorkHours(timestamp: timestamp!, appSettings: appSettings) : false

        if let matchingRule = findMatchingRule(
            tripType: tripType,
            startAddressType: startAddressType,
            endAddressType: endAddressType,
            startAddress: startAddress,
            endAddress: endAddress,
            rules: rules
        ) {
            switch matchingRule.category {
            case "BUSINESS": return .business
            case "COMMUTE": return .commute
            case "PRIVATE": return .privateTrip
            default: return defaultCategory
            }
        }

        if homeWorkAsCommute && (isCommuteAddressPair || tripType == "COMMUTE" || tripType == dynamicsHomeToWork) {
            return .commute
        }
        if customerAsBusiness && (startAddressType == "KLANT" || endAddressType == "KLANT") {
            return .business
        }
        if let type = tripType, businessTripTypes.contains(type) {
            return .business
        }
        if tripType == "PRIVATE" || tripType == dynamicsPersonal {
            return .privateTrip
        }
        
        if outsideHours {
            return .privateTrip
        }
        
        return defaultCategory
    }

    static func resolveTripType(
        tripType: String?,
        startAddressType: String?,
        endAddressType: String?,
        startAddress: String?,
        endAddress: String?,
        rules: [ClassificationRule]
    ) -> String? {
        return findMatchingRule(
            tripType: tripType,
            startAddressType: startAddressType,
            endAddressType: endAddressType,
            startAddress: startAddress,
            endAddress: endAddress,
            rules: rules
        )?.tripType ?? tripType
    }

    private static func findMatchingRule(
        tripType: String?,
        startAddressType: String?,
        endAddressType: String?,
        startAddress: String?,
        endAddress: String?,
        rules: [ClassificationRule]
    ) -> ClassificationRule? {
        let matchedRules = rules.filter { rule in
            let tripMatches = (rule.tripType?.isEmpty ?? true) || rule.tripType?.caseInsensitiveCompare(tripType ?? "") == .orderedSame
            
            let typeDirectionForward = ((rule.startAddressType?.isEmpty ?? true) || rule.startAddressType?.caseInsensitiveCompare(startAddressType ?? "") == .orderedSame) &&
                ((rule.endAddressType?.isEmpty ?? true) || rule.endAddressType?.caseInsensitiveCompare(endAddressType ?? "") == .orderedSame)
                
            let typeDirectionReverse = ((rule.startAddressType?.isEmpty ?? true) || rule.startAddressType?.caseInsensitiveCompare(endAddressType ?? "") == .orderedSame) &&
                ((rule.endAddressType?.isEmpty ?? true) || rule.endAddressType?.caseInsensitiveCompare(startAddressType ?? "") == .orderedSame)

            let exactDirectionForward = (!(rule.startAddress?.isEmpty ?? true) && rule.startAddress?.caseInsensitiveCompare(startAddress ?? "") == .orderedSame) &&
                (!(rule.endAddress?.isEmpty ?? true) && rule.endAddress?.caseInsensitiveCompare(endAddress ?? "") == .orderedSame)
                
            let exactDirectionReverse = (!(rule.startAddress?.isEmpty ?? true) && rule.startAddress?.caseInsensitiveCompare(endAddress ?? "") == .orderedSame) &&
                (!(rule.endAddress?.isEmpty ?? true) && rule.endAddress?.caseInsensitiveCompare(startAddress ?? "") == .orderedSame)
                
            let matchesTypes = typeDirectionForward || typeDirectionReverse
            let matchesExact = exactDirectionForward || exactDirectionReverse
            
            let usesExactAddresses = !(rule.startAddress?.isEmpty ?? true) && !(rule.endAddress?.isEmpty ?? true)
            
            return tripMatches && (usesExactAddresses ? matchesExact : matchesTypes)
        }
        
        return matchedRules.max { a, b in
            let countA = [a.startAddressType, a.endAddressType, a.tripType, a.startAddress, a.endAddress].compactMap { $0 }.filter { !$0.isEmpty }.count
            let countB = [b.startAddressType, b.endAddressType, b.tripType, b.startAddress, b.endAddress].compactMap { $0 }.filter { !$0.isEmpty }.count
            return countA < countB
        }
    }

    static func isOutsideWorkHours(timestamp: Date, appSettings: AppSettings?) -> Bool {
        guard let settings = appSettings else { return false }
        
        let calendar = Calendar.current
        let calDayOfWeek = calendar.component(.weekday, from: timestamp)
        let ourDayOfWeek = calDayOfWeek == 1 ? 7 : calDayOfWeek - 1
        
        let daysStr = settings.workDays
        let daysMap = daysStr.split(separator: ";").compactMap { String($0) }
        
        var isWorkDay = false
        var startMins = 8 * 60
        var endMins = 17 * 60
        
        for dayItem in daysMap {
            let parts = dayItem.split(separator: "=")
            if parts.count == 2, let dayNum = Int(parts[0]), dayNum == ourDayOfWeek {
                isWorkDay = true
                let times = parts[1].split(separator: "-")
                if times.count == 2 {
                    let startParts = times[0].split(separator: ":")
                    let endParts = times[1].split(separator: ":")
                    if startParts.count == 2, let h1 = Int(startParts[0]), let m1 = Int(startParts[1]) {
                        startMins = h1 * 60 + m1
                    }
                    if endParts.count == 2, let h2 = Int(endParts[0]), let m2 = Int(endParts[1]) {
                        endMins = h2 * 60 + m2
                    }
                }
                break
            }
        }
        
        if !isWorkDay {
            return true
        }
        
        let currentHour = calendar.component(.hour, from: timestamp)
        let currentMin = calendar.component(.minute, from: timestamp)
        let currentMins = currentHour * 60 + currentMin
        
        if currentMins < startMins || currentMins > endMins {
            return true
        }
        
        return false
    }
}
import Foundation
import SwiftData

struct PatternDetector {
    /// Detects if the current trip type has been manually selected 3 or more times for this specific route.
    /// If so, returns a proposed `ClassificationRule`.
    @MainActor
    static func detectRuleSuggestion(
        for trip: Trip,
        context: ModelContext,
        allTrips: [Trip]
    ) -> ClassificationRule? {
        guard !(trip.startAddress ?? "").isEmpty, !(trip.endAddress ?? "").isEmpty else { return nil }
        
        let startMatch = AddressMatching.normalizeBase(trip.startAddress ?? "")
        let endMatch = AddressMatching.normalizeBase(trip.endAddress ?? "")
        
        let sameRouteTrips = allTrips.filter { t in
            t.vehicle?.id == trip.vehicle?.id &&
            AddressMatching.normalizeBase(t.startAddress ?? "") == startMatch &&
            AddressMatching.normalizeBase(t.endAddress ?? "") == endMatch &&
            t.tripType == trip.tripType
        }
        
        if sameRouteTrips.count >= 3 {
            // Suggest a rule
            let category: String
            switch TripClassification.classify(tripType: trip.tripType) {
            case .business: category = "BUSINESS"
            case .commute: category = "COMMUTE"
            case .privateTrip: category = "PRIVATE"
            }
            
            return ClassificationRule(
                name: "Auto-suggestie: \(startMatch) -> \(endMatch)",
                startAddress: trip.startAddress ?? "",
                endAddress: trip.endAddress ?? "",
                tripType: trip.tripType,
                category: category
            )
        }
        
        return nil
    }
}
